File: packages/core/src/tim.ts
/**
 * TIM produce + verify, per ARKY-TIM-v1.
 *
 * Canonical body = TIM minus `cid`, `sig`, and `time.witnesses` (the witnesses
 * are co-signed over the same bytes and appended afterwards). The issuer `sig`
 * and each witness are detached-payload Ed25519 JWS over the canonical bytes.
 */

import { canonicalize } from './canonicalize.ts';
import { cidFromCanonical, base58btcDecode } from './cid.ts';
import { signDetached, verifyDetached } from './jws.ts';

export interface TimMeasurement {
  name: string;
  value: number | string | boolean | object;
  unit?: string;
  method: { type: string; source: string; params?: object; version?: string };
  device?: string;
  error?: string;
  code?: string;
  provenance?: unknown;
}

export interface Tim {
  time: { ts: string; witnesses?: string[]; ordering?: unknown };
  identity: { id: string; claims?: unknown[]; proofs?: unknown[] };
  measurement: TimMeasurement;
  prev?: string;
  cid: string;
  nonce?: string;
  exp?: string;
  sig: string;
  [k: string]: unknown;
}

/** The canonical body: strip cid, sig, and time.witnesses (drop empty time). */
export function canonicalBody(tim: Record<string, unknown>): Record<string, unknown> {
  const { cid: _c, sig: _s, ...rest } = tim;
  const time = rest.time as Record<string, unknown> | undefined;
  if (time && 'witnesses' in time) {
    const { witnesses: _w, ...timeRest } = time;
    if (Object.keys(timeRest).length === 0) {
      const { time: _t, ...noTime } = rest;
      return noTime;
    }
    return { ...rest, time: timeRest };
  }
  return rest;
}

export interface CreateTimInput {
  ts: string;
  identity: { id: string; claims?: unknown[]; proofs?: unknown[] };
  measurement: TimMeasurement;
  prev?: string;
  nonce?: string;
  exp?: string;
}

/** Build a signed TIM from inputs and an Ed25519 private key seed. */
export function createTim(input: CreateTimInput, privateKey: Uint8Array, kid?: string): Tim {
  const body: Record<string, unknown> = {
    time: { ts: input.ts },
    identity: input.identity,
    measurement: input.measurement,
    ...(input.prev ? { prev: input.prev } : {}),
    ...(input.nonce ? { nonce: input.nonce } : {}),
    ...(input.exp ? { exp: input.exp } : {}),
  };
  const canonical = canonicalize(body);
  const cid = cidFromCanonical(canonical);
  const sig = signDetached(new TextEncoder().encode(canonical), privateKey, kid);
  return { ...(body as object), cid, sig } as Tim;
}

const REQUIRED_PATHS = [
  'time.ts',
  'identity.id',
  'measurement.name',
  'measurement.value',
  'measurement.method',
  'cid',
  'sig',
];

function getPath(obj: any, path: string): unknown {
  return path.split('.').reduce((a, k) => (a == null ? undefined : a[k]), obj);
}

export interface VerifyResult {
  valid: boolean;
  errors: string[];
  schema_valid: boolean;
  cid_valid: boolean;
  signature_valid: boolean;
  witnesses_valid: boolean;
  /** false iff `options.at` is provided and the TIM's `exp` is at/before it. */
  fresh: boolean;
  missing_fields?: string[];
}

export interface VerifyOptions {
  /**
   * Current time (RFC3339 or epoch ms) for freshness enforcement (TIM §4). When
   * provided, a TIM whose `exp` is at or before this time fails with
   * `tim.expired`. Omit to skip freshness (pure cryptographic verification).
   *
   * Note: anti-replay (`nonce`) and causal-chain (`prev`, cross-identity)
   * enforcement need external state (a seen-nonce store, the prior chain) and
   * are the caller's responsibility — single-TIM verification cannot do them.
   */
  at?: string | number;
}

/**
 * Verify a TIM. `resolveKey` maps an identity/kid to an Ed25519 public key; if
 * omitted, did:key identities are resolved automatically and others fail.
 */
export function verifyTim(
  tim: Record<string, unknown>,
  resolveKey: (tim: Record<string, unknown>) => Uint8Array | undefined = resolveDidKey,
  options: VerifyOptions = {},
): VerifyResult {
  const errors: string[] = [];
  const missing = REQUIRED_PATHS.filter((p) => getPath(tim, p) === undefined);
  const schema_valid = missing.length === 0;
  if (!schema_valid) {
    errors.push('tim.missing_required');
    return {
      valid: false,
      errors,
      schema_valid: false,
      cid_valid: false,
      signature_valid: false,
      witnesses_valid: false,
      fresh: true,
      missing_fields: missing,
    };
  }

  const canonical = canonicalize(canonicalBody(tim));
  const payload = new TextEncoder().encode(canonical);

  const cid_valid = cidFromCanonical(canonical) === tim.cid;
  if (!cid_valid) errors.push('tim.cid_mismatch');

  const key = resolveKey(tim);
  let signature_valid = false;
  if (!key) {
    errors.push('tim.key_unresolved');
  } else {
    signature_valid = verifyDetached(tim.sig as string, payload, key);
    if (!signature_valid) errors.push('tim.invalid_signature');
  }

  // Witnesses (optional): each is a detached JWS over the SAME canonical bytes.
  let witnesses_valid = true;
  const witnesses = (tim.time as any)?.witnesses;
  if (Array.isArray(witnesses)) {
    for (let i = 0; i < witnesses.length; i++) {
      const wk = resolveKey({ ...tim, __witness: witnesses[i] });
      if (!wk || !verifyDetached(witnesses[i], payload, wk)) {
        witnesses_valid = false;
        errors.push(`tim.invalid_witness[${i}]`);
      }
    }
  }

  // Freshness (TIM §4): if a reference time is given and `exp` is at/before it,
  // the receipt has expired.
  let fresh = true;
  if (options.at !== undefined && typeof tim.exp === 'string') {
    const now = typeof options.at === 'number' ? options.at : Date.parse(options.at);
    const exp = Date.parse(tim.exp);
    if (!Number.isNaN(now) && !Number.isNaN(exp) && exp <= now) {
      fresh = false;
      errors.push('tim.expired');
    }
  }

  const valid = cid_valid && signature_valid && witnesses_valid && fresh;
  return { valid, errors, schema_valid, cid_valid, signature_valid, witnesses_valid, fresh };
}

/**
 * Extract an Ed25519 public key from a did:key:z6Mk… identity. Returns
 * undefined for any malformed input (bad base58, wrong multicodec, wrong
 * length) — it MUST NOT throw, so a verifier processing untrusted TIMs cannot
 * be crashed (DoS) by a hostile identity string.
 */
export function resolveDidKey(tim: Record<string, unknown>): Uint8Array | undefined {
  const id = (tim.identity as any)?.id as string | undefined;
  if (!id?.startsWith('did:key:z6Mk')) return undefined;
  try {
    // did:key multibase: 'z' + base58btc(0xed 0x01 || 32-byte ed25519 pubkey).
    const decoded = base58btcDecode(id.slice('did:key:z'.length));
    if (decoded.length !== 34 || decoded[0] !== 0xed || decoded[1] !== 0x01) return undefined;
    return decoded.slice(2);
  } catch {
    return undefined;
  }
}

File: packages/core-rs/src/tim.rs
//! TIM verify, per ARKY-TIM-v1. Canonical body = TIM minus `cid`, `sig`, and
//! `time.witnesses`; issuer `sig` and each witness are detached Ed25519 JWS over
//! those bytes. Public key resolved from a did:key identity (§6.1).

use crate::canonicalize::canonicalize;
use crate::cid::{cid_from_canonical, from_multibase};
use crate::jws::{
    decode_protected_header, sign_detached, verify_detached, verifying_key_from_bytes,
};
use crate::kernel::parse_rfc3339_ms;
use ed25519_dalek::SigningKey;
use serde_json::{Map, Value};

/// Result of verifying a TIM.
#[derive(Debug, Clone)]
pub struct VerifyResult {
    pub valid: bool,
    pub schema_valid: bool,
    pub cid_valid: bool,
    pub signature_valid: bool,
    pub witnesses_valid: bool,
    /// false iff `at` was provided to `verify_tim_at` and the TIM's `exp` is at
    /// or before it (TIM §4). Always true from `verify_tim` (no freshness check).
    pub fresh: bool,
    pub errors: Vec<String>,
    pub missing_fields: Vec<String>,
}

const REQUIRED_PATHS: &[&[&str]] = &[
    &["time", "ts"],
    &["identity", "id"],
    &["measurement", "name"],
    &["measurement", "value"],
    &["measurement", "method"],
    &["cid"],
    &["sig"],
];

fn get_path<'a>(v: &'a Value, path: &[&str]) -> Option<&'a Value> {
    let mut cur = v;
    for k in path {
        cur = cur.get(*k)?;
    }
    Some(cur)
}

/// The canonical body: clone, then remove cid, sig, and time.witnesses (drop an
/// emptied `time`).
pub fn canonical_body(tim: &Value) -> Value {
    let mut obj: Map<String, Value> = tim.as_object().cloned().unwrap_or_default();
    obj.remove("cid");
    obj.remove("sig");
    if let Some(Value::Object(time)) = obj.get_mut("time") {
        time.remove("witnesses");
        if time.is_empty() {
            obj.remove("time");
        }
    }
    Value::Object(obj)
}

/// Build a signed TIM from a body (any JSON object containing time/identity/
/// measurement and optional prev/nonce/exp) and an Ed25519 signing key. Computes
/// the canonical body, cid, and detached `sig`, returning the full TIM.
pub fn create_tim(body: Value, signing_key: &SigningKey, kid: Option<&str>) -> Value {
    let mut obj = body.as_object().cloned().unwrap_or_default();
    obj.remove("cid");
    obj.remove("sig");
    let base = Value::Object(obj);
    let canonical = canonicalize(&base);
    let cid = cid_from_canonical(&canonical);
    let sig = sign_detached(canonical.as_bytes(), signing_key, kid);
    let mut out = base.as_object().unwrap().clone();
    out.insert("cid".into(), Value::String(cid));
    out.insert("sig".into(), Value::String(sig));
    Value::Object(out)
}

/// Extract an Ed25519 public key from a did:key:z6Mk… identity (multicodec
/// 0xed 0x01 || 32-byte key).
pub fn resolve_did_key(id: &str) -> Option<Vec<u8>> {
    let rest = id.strip_prefix("did:key:z")?;
    let decoded = from_multibase(&format!("z{rest}")).ok()?;
    if decoded.len() < 2 || decoded[0] != 0xed || decoded[1] != 0x01 {
        return None;
    }
    Some(decoded[2..].to_vec())
}

/// Resolve a public key for a witness JWS by its `kid` against a small known-key
/// map, falling back to the TIM's did:key identity.
type KeyResolver<'a> = dyn Fn(&Value, Option<&str>) -> Option<Vec<u8>> + 'a;

/// Verify a TIM. `resolve` maps (tim, optional witness-kid) -> 32-byte pubkey.
/// Equivalent to `verify_tim_at(tim, resolve, None)` — a pure cryptographic check
/// with no freshness enforcement (`fresh` is always true).
pub fn verify_tim(tim: &Value, resolve: &KeyResolver) -> VerifyResult {
    verify_tim_at(tim, resolve, None)
}

/// Verify a TIM with an optional reference time for freshness (TIM §4). When `at`
/// is `Some` and the TIM's `exp` parses as an RFC3339 timestamp at or before `at`,
/// the result is `fresh: false` and carries a `tim.expired` error; `valid`
/// includes `fresh` exactly like @arky/core. Omit `at` for a pure cryptographic
/// check. `parse_rfc3339_ms` returns `None` for malformed input (mirrors
/// ECMAScript `Date.parse` NaN), so this path cannot error on bad timestamps.
pub fn verify_tim_at(tim: &Value, resolve: &KeyResolver, at: Option<&str>) -> VerifyResult {
    let mut errors = Vec::new();
    let missing: Vec<String> = REQUIRED_PATHS
        .iter()
        .filter(|p| get_path(tim, p).is_none())
        .map(|p| p.join("."))
        .collect();
    if !missing.is_empty() {
        errors.push("tim.missing_required".to_string());
        return VerifyResult {
            valid: false,
            schema_valid: false,
            cid_valid: false,
            signature_valid: false,
            witnesses_valid: false,
            fresh: true,
            errors,
            missing_fields: missing,
        };
    }

    let canonical = canonicalize(&canonical_body(tim));
    let payload = canonical.as_bytes();

    let stored_cid = tim.get("cid").and_then(|v| v.as_str()).unwrap_or("");
    let cid_valid = cid_from_canonical(&canonical) == stored_cid;
    if !cid_valid {
        errors.push("tim.cid_mismatch".to_string());
    }

    let sig = tim.get("sig").and_then(|v| v.as_str()).unwrap_or("");
    let mut signature_valid = false;
    match resolve(tim, None) {
        Some(kb) => match verifying_key_from_bytes(&kb) {
            Ok(vk) => {
                signature_valid = verify_detached(sig, payload, &vk);
                if !signature_valid {
                    errors.push("tim.invalid_signature".to_string());
                }
            }
            Err(_) => errors.push("tim.key_unresolved".to_string()),
        },
        None => errors.push("tim.key_unresolved".to_string()),
    }

    // Witnesses (optional): each a detached JWS over the SAME canonical bytes.
    let mut witnesses_valid = true;
    if let Some(ws) = get_path(tim, &["time", "witnesses"]).and_then(|v| v.as_array()) {
        for (i, w) in ws.iter().enumerate() {
            let ws_str = w.as_str().unwrap_or("");
            let kid = decode_protected_header(ws_str)
                .ok()
                .and_then(|h| h.get("kid").and_then(|v| v.as_str()).map(String::from));
            let ok = resolve(tim, kid.as_deref())
                .and_then(|kb| verifying_key_from_bytes(&kb).ok())
                .map(|vk| verify_detached(ws_str, payload, &vk))
                .unwrap_or(false);
            if !ok {
                witnesses_valid = false;
                errors.push(format!("tim.invalid_witness[{i}]"));
            }
        }
    }

    // Freshness (TIM §4): if a reference time is given and `exp` is at/before it,
    // the receipt has expired. Mirrors @arky/core: only string `exp` is checked,
    // and unparseable `at`/`exp` (None) are ignored rather than erroring.
    let mut fresh = true;
    if let Some(at_str) = at
        && let Some(exp_str) = tim.get("exp").and_then(|v| v.as_str())
        && let (Some(now), Some(exp)) = (parse_rfc3339_ms(at_str), parse_rfc3339_ms(exp_str))
        && exp <= now
    {
        fresh = false;
        errors.push("tim.expired".to_string());
    }

    let valid = cid_valid && signature_valid && witnesses_valid && fresh;
    VerifyResult {
        valid,
        schema_valid: true,
        cid_valid,
        signature_valid,
        witnesses_valid,
        fresh,
        errors,
        missing_fields: vec![],
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::keys::from_seed;
    use serde_json::json;

    /// Default resolver: resolve the TIM identity did:key, ignore witness kid.
    fn identity_resolver(tim: &Value, _kid: Option<&str>) -> Option<Vec<u8>> {
        let id = tim.get("identity")?.get("id")?.as_str()?;
        resolve_did_key(id)
    }

    #[test]
    fn freshness_expired_with_at_unexpired_without() {
        let kp = from_seed([7u8; 32]);
        let body = json!({
            "time": { "ts": "2025-10-15T12:00:00Z" },
            "identity": { "id": kp.did },
            "measurement": {
                "name": "temp", "value": 22.5, "unit": "degC",
                "method": { "type": "sensor", "source": "s" }
            },
            "exp": "2020-01-02T00:00:00Z",
        });
        let tim = create_tim(body, &kp.signing_key, None);

        // No `at`: pure cryptographic check — fresh stays true, valid true.
        let r = verify_tim(&tim, &identity_resolver);
        assert!(r.fresh, "fresh should be true without at");
        assert!(r.valid, "valid should be true without at");

        // With `at` after exp: fresh false, valid false, tim.expired error.
        let r = verify_tim_at(&tim, &identity_resolver, Some("2026-01-01T00:00:00Z"));
        assert!(!r.fresh, "fresh should be false when expired");
        assert!(!r.valid, "valid should be false when expired");
        assert!(r.errors.iter().any(|e| e == "tim.expired"));
    }

    #[test]
    fn freshness_future_exp_passes_with_at() {
        let kp = from_seed([8u8; 32]);
        let body = json!({
            "time": { "ts": "2025-10-15T12:00:00Z" },
            "identity": { "id": kp.did },
            "measurement": {
                "name": "temp", "value": 22.5, "unit": "degC",
                "method": { "type": "sensor", "source": "s" }
            },
            "exp": "2099-01-01T00:00:00Z",
        });
        let tim = create_tim(body, &kp.signing_key, None);
        let r = verify_tim_at(&tim, &identity_resolver, Some("2026-01-01T00:00:00Z"));
        assert!(r.fresh);
        assert!(r.valid);
    }
}
